### PR TITLE
[WIP] Upgrade to react 16.2.0

### PR DESCRIPTION
**Pull request by AnshulMalik**
Associated Issue: #4879 

### Summary of Changes

* React and React DOM to 16.2.0

### CODE REVIEW

**Comment by jasonLaster** [comment]
Thanks @AnshulMalik could you update the jest tests.

i can look into these license

**Comment by AnshulMalik** [comment]
Yep, working on the tests :)

**Comment by darkwing** [comment]
Hey @AnshulMalik !  Thank you for this!

Could you add a checklist to the description so we can track the remaining WIP items?  It's something I'd like to try out to keep the PR queue "clean" moving forward.  Thank you!

**Comment by jasonLaster** [comment]
^ @nchevobbe did we address this in core?

**Comment by nchevobbe** [comment]
just run license-check on devtools-core without issues.
I guess we should evaluate those dependencies and their licences

File: src/components/Editor/tests/Editor.spec.js [hunk 1]
```diff
@@ -117,7 +117,6 @@ describe("Editor", () => {
       });
 
       expect(mockEditor.setText.mock.calls).toEqual([
-        [""],
```
Comment (AnshulMalik) [inline src/components/Editor/tests/Editor.spec.js]: Is it okay that this one call is being skipped?

**Comment by jasonLaster** [comment]
I'm pleasantly surprised that we don't pull in react 15.x

File: src/components/Editor/tests/Editor.spec.js (hunk 1 above)
Comment (jasonLaster) [reply src/components/Editor/tests/Editor.spec.js]: yep

File: src/components/tests/__snapshots__/QuickOpenModal.spec.js.snap [hunk 2]
```diff
@@ -1458,14 +1458,14 @@ exports[`QuickOpenModal updateResults on enable 2`] = `
     >
       <Modal
         handleClose={[Function]}
-        status="entering"
+        status="exited"
```
Comment (nchevobbe) [inline src/components/tests/__snapshots__/QuickOpenModal.spec.js.snap]: is this intended ?

Comment (jasonLaster) [reply src/components/tests/__snapshots__/QuickOpenModal.spec.js.snap]: yeah - i'm scared of that too... i feel like we have that problem in the panel perhaps, so we should land this and then test that...

Comment (jasonLaster) [reply src/components/tests/__snapshots__/QuickOpenModal.spec.js.snap]: i guess i could also test it now... before it lands :wink:

Comment (nchevobbe) [reply src/components/tests/__snapshots__/QuickOpenModal.spec.js.snap]: yeah, i think it would be safer :)

**Comment by darkwing** [comment]
When toggling the right pane, I see: "Warning: Expressions has a method called shouldComponentUpdate(). shouldComponentUpdate should not be used when extending React.PureComponent. Please extend React.Component if shouldComponentUpdate is used"

That doesn't appear to happen on master.

**Comment by darkwing** [comment]
`Warning: EditorMenu has a method called shouldComponentUpdate(). shouldComponentUpdate should not be used when extending React.PureComponent. Please extend React.Component if shouldComponentUpdate is used.`

It looks like our PureComponent usage is wrong for this upgrade.

File: src/components/tests/__snapshots__/QuickOpenModal.spec.js.snap (hunk 2 above)
Comment (jasonLaster) [inline src/components/tests/__snapshots__/QuickOpenModal.spec.js.snap]: Quick open slides in and slides out. It is also fine in the other 10 unit tests. 

I am going to move forward and we can figure this mystery out later. I believe it is just a timing thing.

**Comment by jasonLaster** [comment]
Same w/ the jest tests. I think this can also be a follow up as it is a warning, but not an error:

**Comment by jasonLaster** [comment]
The necessity for moving forward is that we need to be on the same version of react as the panel to catch bugs that only come up between the differences.

**Pull request merged**